### DISCUSSION TITLE
Issue with "Pin" object

### DISCUSSION THREAD

**Discussion by mmliam** [micropython/micropython#11158] [ESP8266]
Hi,

Needed to flip the ESP8266 (ESP-07s module) pin 2 from output to input.
Using this version of micropython: v1.191(583K)

This is the example on micropython.org: (https://docs.micropython.org/en/latest/library/machine.Pin.html)
# reconfigure pin #0 in input mode with a pull down resistor
p0.init(p0.IN, p0.PULL_DOWN)

Used the following instruction for my pin object: pin_2.init(pin_2.IN)
The pin remained an output; it did not change to an input.

Changed the instruction to this: pin_2(pin_2.IN).
This format changed the pin from output to input as needed.

Is the example on micropython.org wrong, or is there some other issue?

mm

**Comment by robert-hh** [comment]
You did not tell which board you use. Some boards like the Wemos D1 Mini have a pull-up resistor at Pin 2. So even when set to Input, pull-down, the level will be not 0. 
pin_2(pin_2.IN) is the short form of pin_2.value(pin_2.IN). Since pin2.IN has the value of 0, it will set the output to 0.

**Reply by mmliam** [reply]
Hi robert-hh,
I did specify that I was using the ESP-07s module.
Regarding the instruction pin_2(pin_2.IN) I clearly misunderstood it's function.
But, when I used pin_2.init(pin_2.IN), I expected the voltage to drop to zero, it only dropped about 5%. Apparently there was still a pull-up resistor at what was now the inputl even though I set pull=None.

This is how I initialize the pin object:
pin_2 = Pin(2, Pin.OUT, pull=None, value=highState) NOTE: highState = 1

Shouldn't pull=None eliminate pull-up and pull-down resistors?

**Reply by robert-hh** [reply]
The pull setting would only apply to IN mode. For output mode, it does not matter. The internal pull resistor is ~30k, which will be overruled by the active output driver. You have to specify pull=xx when setting the input mode. I do not know how you determined the 5% drop.

**Reply by mmliam** [reply]
Amazing, I was just going to try the "pull" for the input - lol
I wish I could find a reliable schematic of at least the I/O section of the ESP8266.
If you don't specify an open-collector is the output a tri-state?

> I do not know how you determined the 5% drop.

I measured the voltage at pin 2 with a multimeter; I put strategic sleep() statements in the code to allow the meter to stabilize.
I could also have used this amazing portable combination multimeter/scope I recently purchased on Amazon:
https://www.amazon.com/dp/B08VRDG94N?psc=1&ref=ppx_yo2ov_dt_b_product_details
It's well worth the $126.


UPDATE: Problem solved: I switched to pin 5, voltage now drops to zero when configured as an input.
I could kick myself; pin 2 on the ESP-07s is also connected to the LED. That's why the voltage only dropped to 5% when switching pin 2 to an input. 
GOOD GRIEF.

Thanks for your help robert.
